### PR TITLE
First version of a merge bot.

### PR DESCRIPTION
**Pull request by SirVer**

### CODE REVIEW

File: wally_the_cartographer/cmd/root.go
Comment (cschuet) [inline wally_the_cartographer/cmd/root.go]: its

File: wally_the_cartographer/cmd/root.go
Comment (cschuet) [inline wally_the_cartographer/cmd/root.go]: Is there nothing more useful that you can say about RootCmd?

File: wally_the_cartographer/cmd/root.go
Comment (cschuet) [inline wally_the_cartographer/cmd/root.go]: Say something more useful.

**Comment by ojura** [comment]
@SirVer, just to respond to your [comment](https://github.com/googlecartographer/cartographer/issues/649#issuecomment-343446937):

> rebasing makes reviewing easier in case further manual changes are required or even when the bot adds more commits of its own (clang-format and so on)

I don't think it makes a big difference in reviewing or making further additional commits.
For reviewing, you would just review the diff from master to the merge commit (which is what Github already does for you when you press `Files changed` at the top of the Github UI on the PR page). How the tree looks doesn't really matter there.
Further additional commits can always be added on top of the merge commit just as easily as on top of the rebased commits.

Personally, when updating a PR manually, I always prefer a rebase, so I understand you. But sometimes the PR grows (not neccessarily in lines changed, but in number of commits) and rebasing would mean that you would have to resolve basically the same conflict many times. On the other hand, merging master allows you do to this by resolving conflicts only once, in the merge commit. And, there is a chance that you can even avoid some conflicts, as described.

There is also another possibility: squash-then-rebase. In the scenario above, squashing `C -> revertC -> D` would effectively yield a commit that is identical to `D`, which could easily be rebased on top of `B`. So you get the desired effect of avoiding conflicts as described, while still doing a rebase.

But still, I think I have one additional convincing argument against rebasing: since you're force-pushing rewritten commits into someone else's private branch, this prevents them from pulling these changes. It's considered rude behaviour to do this when using git.

**Comment by SirVer** [comment]
@ojura most of your comment is about reviewing style. I for one find the merge commit in the review history to be distracting and diluting the diff of the branch, since i mostly review by commit. 

> Also, I have one additional argument against rebasing: since you're force-pushing rewritten commits into someone else's private branch, this prevents them from pulling these changes. It's considered rude behaviour to do this when using git.

Not sure about the rudeness of this. It is typical if two or more people directly collaborate on a branch that you need to `git pull --force` from time to time. I have been doing this in open source for years without much friction. However, I agree that it is unfortunate that we have to push to the PR branch - unfortunately that is a limitation of GitHub: Since a PR is only marked as "merged" if either 

1. it is not rebased and merged (so that a SHA matches) or
2. the "merge button" or API.

Therefore we cannot squash & merge in the fork of the bot ourself, but instead have to use the API to merge. This means that the bot needs to push to the review branch. 

Your concerns about rebasing vs merging are heard. I want to test it out for a bit and see if it is annoying to contributors. We can always change it to merging later on.

**Comment by ojura** [comment]
Sure! Also, I edited in one additional paragraph above that you may have missed:

> There is also another possibility: squash-then-rebase. In the scenario above, squashing C -> revertC -> D would effectively yield a commit that is identical to D, which could easily be rebased on top of B. So you get the desired effect of avoiding conflicts as described, while still doing a rebase.

File: wally_the_cartographer/cmd/root.go
Comment (SirVer) [reply wally_the_cartographer/cmd/root.go]: no. I really don't quite understand how Cobra works.

File: wally_the_cartographer/cmd/root.go
Comment (SirVer) [reply wally_the_cartographer/cmd/root.go]: :man_shrugging:  Sorry

**Comment by ojura** [comment]
> Wally B. Feed was a cartographer on Scabb Island.
  He claimed to have traveled all the world, and putting every place he'd seen on paper. As well as making maps, he did restoration work on them, repairs, copying and 'artsy fartsy stuff' like painting little cupids in the corners.

Ayy, props to whoever came up with the idea for the bot's name!

**Comment by cschuet** [comment]
That'd be @SirVer.

**Review by cschuet — APPROVED** [review]

**Comment by SirVer** [comment]
Gonna merge, though I expect a bit more bug fixes coming down today.

@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Error: PUT https://api.github.com/repos/googlecartographer/infrastructure/pulls/1/merge: 404 Not Found []

Log:
~~~
=> Rebasing PR 1 on googlecartographer/infrastructure onto master.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/infrastructure]
Already up to date.
 $ git remote add SirVer git@github.com:SirVer/infrastructure.git [data/infrastructure]
 $ git fetch SirVer [data/infrastructure]
From github.com:SirVer/infrastructure
 * [new branch]      master     -> SirVer/master
 * [new branch]      wally      -> SirVer/wally
 $ git branch --track pr_1 SirVer/wally [data/infrastructure]
Branch 'pr_1' set up to track remote branch 'wally' from 'SirVer'.
=> Switching to branch pr_1.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout pr_1 [data/infrastructure]
Switched to branch 'pr_1'
Your branch is up to date with 'SirVer/wally'.
 $ git pull --force [data/infrastructure]
Already up to date.
 $ git rebase master [data/infrastructure]
Current branch pr_1 is up to date.
 $ git push --force SirVer HEAD:wally [data/infrastructure]
Everything up-to-date
=> Abandoning branch pr_1.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'SirVer/wally'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1 [data/infrastructure]
Deleted branch pr_1 (was aed100a).

~~~

**Comment by wally-the-cartographer** [comment]
Error: PUT https://api.github.com/repos/googlecartographer/infrastructure/pulls/1/merge: 404 Not Found []

Log:
~~~
=> Rebasing PR 1 on googlecartographer/infrastructure onto master.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/infrastructure]
Already up to date.
 $ git remote add SirVer git@github.com:SirVer/infrastructure.git [data/infrastructure]
fatal: remote SirVer already exists.
 $ git fetch SirVer [data/infrastructure]
 $ git branch --track pr_1 SirVer/wally [data/infrastructure]
Branch 'pr_1' set up to track remote branch 'wally' from 'SirVer'.
=> Switching to branch pr_1.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout pr_1 [data/infrastructure]
Switched to branch 'pr_1'
Your branch is up to date with 'SirVer/wally'.
 $ git pull --force [data/infrastructure]
Already up to date.
 $ git rebase master [data/infrastructure]
Current branch pr_1 is up to date.
 $ git push --force SirVer HEAD:wally [data/infrastructure]
Everything up-to-date
=> Abandoning branch pr_1.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'SirVer/wally'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1 [data/infrastructure]
Deleted branch pr_1 (was aed100a).

~~~

**Comment by wally-the-cartographer** [comment]
Error: PUT https://api.github.com/repos/googlecartographer/infrastructure/pulls/1/merge: 404 Not Found []

Log:
~~~
=> Rebasing PR 1 on googlecartographer/infrastructure onto master.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/infrastructure]
Already up to date.
 $ git remote add SirVer git@github.com:SirVer/infrastructure.git [data/infrastructure]
fatal: remote SirVer already exists.
 $ git fetch SirVer [data/infrastructure]
 $ git branch --track pr_1 SirVer/wally [data/infrastructure]
Branch 'pr_1' set up to track remote branch 'wally' from 'SirVer'.
=> Switching to branch pr_1.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout pr_1 [data/infrastructure]
Switched to branch 'pr_1'
Your branch is up to date with 'SirVer/wally'.
 $ git pull --force [data/infrastructure]
Already up to date.
 $ git rebase master [data/infrastructure]
Current branch pr_1 is up to date.
 $ git push --force SirVer HEAD:wally [data/infrastructure]
Everything up-to-date
=> Abandoning branch pr_1.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'SirVer/wally'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1 [data/infrastructure]
Deleted branch pr_1 (was aed100a).

~~~

**Comment by wally-the-cartographer** [comment]
Error: PUT https://api.github.com/repos/googlecartographer/infrastructure/pulls/1/merge: 404 Not Found []

Log:
~~~
=> Rebasing PR 1 on googlecartographer/infrastructure onto master.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/infrastructure]
Already up to date.
 $ git remote add SirVer git@github.com:SirVer/infrastructure.git [data/infrastructure]
fatal: remote SirVer already exists.
 $ git fetch SirVer [data/infrastructure]
 $ git branch --track pr_1 SirVer/wally [data/infrastructure]
Branch 'pr_1' set up to track remote branch 'wally' from 'SirVer'.
=> Switching to branch pr_1.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout pr_1 [data/infrastructure]
Switched to branch 'pr_1'
Your branch is up to date with 'SirVer/wally'.
 $ git pull --force [data/infrastructure]
Already up to date.
 $ git rebase master [data/infrastructure]
Current branch pr_1 is up to date.
 $ git push --force SirVer HEAD:wally [data/infrastructure]
Everything up-to-date
=> Abandoning branch pr_1.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'SirVer/wally'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1 [data/infrastructure]
Deleted branch pr_1 (was aed100a).

~~~

**Comment by wally-the-cartographer** [comment]
Error: PUT https://api.github.com/repos/googlecartographer/infrastructure/pulls/1/merge: 404 Not Found []

Log:
~~~
=> Rebasing PR 1 on googlecartographer/infrastructure onto master.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/infrastructure]
Already up to date.
 $ git remote add SirVer git@github.com:SirVer/infrastructure.git [data/infrastructure]
fatal: remote SirVer already exists.
 $ git fetch SirVer [data/infrastructure]
 $ git branch --track pr_1 SirVer/wally [data/infrastructure]
Branch 'pr_1' set up to track remote branch 'wally' from 'SirVer'.
=> Switching to branch pr_1.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'origin/master'.
 $ git clean -fd [data/infrastructure]
 $ git checkout pr_1 [data/infrastructure]
Switched to branch 'pr_1'
Your branch is up to date with 'SirVer/wally'.
 $ git pull --force [data/infrastructure]
Already up to date.
 $ git rebase master [data/infrastructure]
Current branch pr_1 is up to date.
 $ git push --force SirVer HEAD:wally [data/infrastructure]
Everything up-to-date
=> Abandoning branch pr_1.
=> Switching to branch master.
 $ git rebase --abort [data/infrastructure]
No rebase in progress?
 $ git checkout -f [data/infrastructure]
Your branch is up to date with 'SirVer/wally'.
 $ git clean -fd [data/infrastructure]
 $ git checkout master [data/infrastructure]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_1 [data/infrastructure]
Deleted branch pr_1 (was aed100a).

~~~

**Comment by SirVer** [comment]
Wally didn't have the rights to merge in this repo.

@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Pull request merged**
